File: win-prog/ProgWinVB/Beziers and Other Splines/Bezier/Bezier.vb
'---------------------------------------
' Bezier.vb (c) 2002 by Charles Petzold
'---------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class Bezier
    Inherits Form

    Protected apt(3) As Point

    Shared Sub Main()
        Application.Run(New Bezier())
    End Sub

    Sub New()
        Text = "Bezier (Mouse Defines Control Points)"
        BackColor = SystemColors.Window
        ForeColor = SystemColors.WindowText
        ResizeRedraw = True

        OnResize(EventArgs.Empty)
    End Sub

    Protected Overrides Sub OnResize(ByVal ea As EventArgs)
        MyBase.OnResize(ea)

        Dim cx As Integer = ClientSize.Width
        Dim cy As Integer = ClientSize.Height

        apt(0) = New Point(cx \ 4, cy \ 2)
        apt(1) = New Point(cx \ 2, cy \ 4)
        apt(2) = New Point(cx \ 2, 3 * cy \ 4)
        apt(3) = New Point(3 * cx \ 4, cy \ 2)
    End Sub

    Protected Overrides Sub OnMouseDown(ByVal mea As MouseEventArgs)
        Dim pt As Point

        If mea.Button = MouseButtons.Left Then
            pt = apt(1)
        ElseIf mea.Button = MouseButtons.Right Then
            pt = apt(2)
        Else
            Return
        End If

        Cursor.Position = PointToScreen(pt)
    End Sub

    Protected Overrides Sub OnMouseMove(ByVal mea As MouseEventArgs)
        If mea.Button = MouseButtons.Left Then
            apt(1) = New Point(mea.X, mea.Y)
            Invalidate()
        ElseIf mea.Button = MouseButtons.Right Then
            apt(2) = New Point(mea.X, mea.Y)
            Invalidate()
        End If
    End Sub

    Protected Overrides Sub OnPaint(ByVal pea As PaintEventArgs)
        Dim grfx As Graphics = pea.Graphics
        grfx.DrawBeziers(New Pen(ForeColor), apt)

        Dim pn As New Pen(Color.FromArgb(128, ForeColor))
        grfx.DrawLine(pn, apt(0), apt(1))
        grfx.DrawLine(pn, apt(2), apt(3))
    End Sub
End Class

File: win-prog/ProgWinCS/Beziers and Other Splines/Bezier/Bezier.cs
//-------------------------------------
// Bezier.cs © 2001 by Charles Petzold
//-------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class Bezier: Form
{
     protected Point[] apt = new Point[4];

     public static void Main()
     {
          Application.Run(new Bezier());
     }
     public Bezier()
     {
          Text = "Bezier (Mouse Defines Control Points)";
          BackColor = SystemColors.Window;
          ForeColor = SystemColors.WindowText;
          ResizeRedraw = true;

		  OnResize(EventArgs.Empty);
     }
     protected override void OnResize(EventArgs ea)
     {
          base.OnResize(ea);

          int cx = ClientSize.Width;
          int cy = ClientSize.Height;

          apt[0] = new Point(    cx / 4,     cy / 2);
          apt[1] = new Point(    cx / 2,     cy / 4);
          apt[2] = new Point(    cx / 2, 3 * cy / 4);
          apt[3] = new Point(3 * cx / 4,     cy / 2);
     }
     protected override void OnMouseDown(MouseEventArgs mea)
     {
          Point pt;

          if (mea.Button == MouseButtons.Left)
               pt = apt[1];

          else if (mea.Button == MouseButtons.Right)
               pt = apt[2];

          else
               return;

          Cursor.Position = PointToScreen(pt);
     }
     protected override void OnMouseMove(MouseEventArgs mea)
     {
          if (mea.Button == MouseButtons.Left)
          {
               apt[1] = new Point(mea.X, mea.Y);
               Invalidate();
          }
          else if (mea.Button == MouseButtons.Right)
          {
               apt[2] = new Point(mea.X, mea.Y);
               Invalidate();
          }
     }
     protected override void OnPaint(PaintEventArgs pea)
     {
          Graphics grfx = pea.Graphics;

          grfx.DrawBeziers(new Pen(ForeColor), apt);

          Pen pen = new Pen(Color.FromArgb(0x80, ForeColor));

          grfx.DrawLine(pen, apt[0], apt[1]);
          grfx.DrawLine(pen, apt[2], apt[3]);
     }
}
